### DISCUSSION TITLE
😭make: *** [Makefile:499: build-info.h] Error 2

### DISCUSSION THREAD

**Discussion by satelliter** [ggml-org/llama.cpp#3052] [Q&A] [answered]
I use 
``` make LLAMA_CUBLAS=1``` 
and it output:
``` I llama.cpp build info: 
I UNAME_S:  Linux
I UNAME_P:  x86_64
I UNAME_M:  x86_64
I CFLAGS:   -I. -Icommon -DNDEBUG -DGGML_USE_K_QUANTS -DGGML_USE_CUBLAS -I/usr/local/cuda/include -I/opt/cuda/include -I/targets/x86_64-linux/include  -O3 -std=c11   -fPIC -Wall -Wextra -Wpedantic -Wcast-qual -Wdouble-promotion -Wshadow -Wstrict-prototypes -Wpointer-arith -Wmissing-prototypes -Werror=implicit-int -Wno-unused-function -pthread -march=native -mtune=native 
I CXXFLAGS: -I. -Icommon -DNDEBUG -DGGML_USE_K_QUANTS -DGGML_USE_CUBLAS -I/usr/local/cuda/include -I/opt/cuda/include -I/targets/x86_64-linux/include  -O3 -std=c++11 -fPIC -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wno-multichar -Wno-format-truncation -pthread -march=native -mtune=native 
I LDFLAGS:   -lcublas -lculibos -lcudart -lcublasLt -lpthread -ldl -lrt -L/usr/local/cuda/lib64 -L/opt/cuda/lib64 -L/targets/x86_64-linux/lib 
I CC:       cc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
I CXX:      g++ (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0

: not foundld-info.sh: 2: 
: not foundld-info.sh: 5: 
scripts/build-info.sh: 24: Syntax error: end of file unexpected (expecting "then")
make: *** [Makefile:499: build-info.h] Error 2
``` 

here is my setup and environment  info:

GPU A100 
``` 
nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2021 NVIDIA Corporation
Built on Sun_Mar_21_19:15:46_PDT_2021
Cuda compilation tools, release 11.3, V11.3.58
Build cuda_11.3.r11.3/compiler.29745058_0
``` 

``` 
git log
commit 178b1850ebd21b349cebbee887950e435c5aa2d3 (HEAD -> master, tag: b1187, origin/master, origin/HEAD)
Author: Georgi Gerganov <ggerganov@gmail.com>
Date:   Wed Sep 6 12:40:57 2023 +0300

    k-quants : fix zero-weight guard in Q6_K (ref #3040)
``` 

Python 3.10.12

```
pip list | egrep "torch|numpy|sentencepiece"

   numpy         1.24.0
  sentencepiece 0.1.98
``` 
``` 
GNU Make 4.2.1
``` 

``` 
lscpu
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          16
On-line CPU(s) list:             0-15
Thread(s) per core:              2
Core(s) per socket:              8
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           106
Model name:                      Intel(R) Xeon(R) Platinum 8369B CPU @ 2.90GHz
Stepping:                        6
CPU MHz:                         2899.998
BogoMIPS:                        5799.99
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       384 KiB
L1i cache:                       256 KiB
L2 cache:                        10 MiB
L3 cache:                        48 MiB
NUMA node0 CPU(s):               0-15
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq monitor ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave
                                  avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves
                                  wbnoinvd arat avx512vbmi avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid arch_capabilities
```

**Comment by BarfingLemurs** [comment, marked as answer]
If make is down, use `cmake`

```
mkdir build
cd build
cmake .. -DLLAMA_CUBLAS=ON
cmake --build . --config Release
```

**Reply by satelliter** [reply]
Thank you very much！ It works.🥰

**Comment by DannyDaemonic** [comment]
Hi @satelliter. The symptoms you've described point to a line-ending mismatch. This often happens when a file with Windows-style line endings (CRLF) is executed on a Unix or Linux system, which expects Unix-style line endings (LF).

#### Breakdown of Symptoms

1. **`: not foundld-info.sh: 2:`**: This is likely due to the shell looking for the Unix line ending (`\n`) but encountering a Windows line ending (`\r\n`). The rogue `\r` character can cause the cursor to move to the beginning of the line, thus overwriting the error message.

2. **`Syntax error: end of file unexpected (expecting "then")`**: This error is probably due to the `if [ $? -eq 0 ]; then` line not being recognized correctly because of incorrect line endings.

3. **Errors on blank lines 2 and 5**: These lines might appear blank but actually contain a `\r`, causing the shell to choke.

#### Potential Causes
1. **Downloading Source Zip**: If you've downloaded the code as a zip archive, some software might automatically convert line endings to Windows-style.
  
2. **Cross-Platform Git Usage**: If you've cloned the repo using Git on a Windows machine and then transferred the code to a Unix/Linux machine without changing the line endings, this issue could arise.

3. **Distro-Specific Issues**: Some Linux distributions or shell configurations may have different behavior for line endings.

#### Solutions

1. **Convert Line Endings**: Use `dos2unix` to convert the line endings of `build-info.sh`:
    ```
    dos2unix build-info.sh
    ```
    Or use `sed` to remove the carriage returns:
    ```
    sed -i 's/\r//' build-info.sh
    ```

2. **Use a Different Editor**: Edit the script in an editor that allows you to specify the line endings, like Vim or Notepad++, and save it with Unix line endings.

3. **Continue Using CMake**: As you've mentioned that CMake works without this issue, continuing to use CMake is an acceptable workaround.

#### Additional Information Needed

To further assist you and prevent similar issues in the future, could you provide us with more information?

1. How did you obtain the code? Did you download a zip archive, use `git clone`, or some other method?
  
2. Did you originally clone or download the code on a different operating system before transferring it to your current machine? If so, which OSes were involved?
  
3. What system and distribution are you using?

4. If you are using Git, could you let us know the output of `git config core.autocrlf`? This will help us understand how Git is handling line endings on your system.

Your answers could help us figure out what happened more precisely and take steps to prevent it for other users.

**Reply by andrey-savov** [reply]
```bash
dos2unix scripts/*
```

**Reply by mcKirkaRo** [reply]
This error appeared when copying the llama.cpp directory from the Windows system in Linux. 
When cloning from the repository, such a thing did not occur.
